ams/server/src/main/java/com/netease/arctic/server/table/executor/DataExpiringExecutor.java

```diff
@@ -0,0 +1,468 @@
+
+    purgeTableData(table, expirationConfig, expireTimestamp);
+  }
+
+  protected static ZoneId getDefaultZoneId(Types.NestedField expireField) {
+    Type type = expireField.type();
+    if (type.typeId() == Type.TypeID.STRING) {
+      return ZoneId.systemDefault();
+    }
+    return ZoneOffset.UTC;
+  }
+
+  private static TableEntriesScan fileEntriesScan(UnkeyedTable table, Expression partitionFilter, Snapshot snapshot) {
+    TableEntriesScan.Builder builder = TableEntriesScan.builder(table)
+        .includeFileContent(FileContent.DATA, FileContent.EQUALITY_DELETES, FileContent.POSITION_DELETES)
+        .withDataFilter(partitionFilter)
+        .includeColumnStats();
+
+    if (null == snapshot || snapshot.snapshotId() == ArcticServiceConstants.INVALID_SNAPSHOT_ID) {
+      return builder.build();
+    } else {
+      return builder.useSnapshot(snapshot.snapshotId()).build();
+    }
+  }
+
+  private static void purgeTableData(ArcticTable table, DataExpirationConfig expirationConfig, long expireTimestamp) {
+    Expression partitionFilter = getPartitionExpression(expirationConfig, expireTimestamp);
+    Map<StructLike, DataFileFreshness> partitionFreshness = Maps.newHashMap();
+
+    if (table.isKeyedTable()) {
+      KeyedTable keyedTable = table.asKeyedTable();
+      ChangeTable changeTable = keyedTable.changeTable();
+      BaseTable baseTable = keyedTable.baseTable();
+      Snapshot changeSnapshot = changeTable.currentSnapshot();
+      Snapshot baseSnapshot = baseTable.currentSnapshot();
+
+      TableEntriesScan changeScan = fileEntriesScan(changeTable, partitionFilter, changeSnapshot);
+      TableEntriesScan baseScan = fileEntriesScan(baseTable, partitionFilter, baseSnapshot);
+      ExpireFiles changeExpiredFiles = new ExpireFiles();
+      ExpireFiles baseExpiredFiles = new ExpireFiles();
+      CloseableIterable<FileEntry> changed = CloseableIterable.transform(changeScan.entries(),
+          e -> new FileEntry(e, true));
+      CloseableIterable<FileEntry> based = CloseableIterable.transform(baseScan.entries(),
+          e -> new FileEntry(e, false));
+
+      try (CloseableIterable<FileEntry> entries = CloseableIterable.withNoopClose(Iterables.concat(changed, based))) {
+        CloseableIterable<FileEntry> mayExpiredFiles = CloseableIterable.withNoopClose(
+            Lists.newArrayList(CloseableIterable.filter(entries,
+                e -> mayExpired(e, expirationConfig, partitionFreshness, expireTimestamp))));
+        CloseableIterable.filter(mayExpiredFiles, e -> willNotRetain(e, expirationConfig, partitionFreshness))
+            .forEach(e -> {
+              if (e.isChange) {
+                changeExpiredFiles.addFile(e);
+              } else {
+                baseExpiredFiles.addFile(e);
+              }
+            });
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      expireFiles(changeTable, changeSnapshot.snapshotId(), changeExpiredFiles, expireTimestamp);
```

Comment:
Is it necessary to set an expiry on the change store for Mixed format, or can it be resolved through the retention time of the change store itself? @zhoujinsong

Reply:
As far as I understand, ChangeStore already has a TTL mechanism based on commit time, but this data expiration feature is a more flexible expiration strategy based on business fields. Both can coexist.